Freshpaint/Classes/FPFreshpaintIntegration.m

```diff
@@ -391,29 +391,46 @@ - (void)notifyForName:(NSString *)name userInfo:(id)userInfo
 
 - (void)sendData:(NSArray *)batch
 {
+    NSString *sessionParameter = [self.analytics validatedSessionId];
+
+    NSMutableArray *mutBatch = [NSMutableArray arrayWithCapacity:batch.count];
+
+    for (NSDictionary *event in batch) {
+        NSMutableDictionary *mutEvent =
+            [NSMutableDictionary dictionaryWithDictionary:event];
+
+        NSMutableDictionary *props =
+            [mutEvent[@"properties"] mutableCopy] ?: [NSMutableDictionary dictionary];
+
+        props[@"$session_id"] = sessionParameter; 
+
+        mutEvent[@"properties"] = [props copy];
+
+        [mutBatch addObject:[mutEvent copy]];
+    }
+
     NSMutableDictionary *payload = [[NSMutableDictionary alloc] init];
-    [payload setObject:iso8601FormattedString([NSDate date]) forKey:@"sentAt"];
-    [payload setObject:batch forKey:@"batch"];
+    payload[@"sentAt"] = iso8601FormattedString([NSDate date]);
+    payload[@"batch"]  = mutBatch;
+
+    FPLog(@"Payload ➜ %@", payload);
 
-    FPLog(@"%@ Flushing %lu of %lu queued API calls.", self, (unsigned long)batch.count, (unsigned long)self.queue.count);
-    FPLog(@"Flushing batch %@.", payload);
+    self.batchRequest =
+        [self.httpClient upload:payload
+                   forWriteKey:self.configuration.writeKey
+            completionHandler:^(BOOL retry) {
 
-    self.batchRequest = [self.httpClient upload:payload forWriteKey:self.configuration.writeKey completionHandler:^(BOOL retry) {
         void (^completion)(void) = ^{
             if (retry) {
                 [self notifyForName:FPFreshpaintRequestDidFailNotification userInfo:batch];
-                self.batchRequest = nil;
-                [self endBackgroundTask];
-                return;
+            } else {
+                [self.queue removeObjectsInArray:batch];
+                [self persistQueue];
+                [self notifyForName:FPFreshpaintRequestDidSucceedNotification userInfo:batch];
```

Comment:
it looks like the branching logic was changed here. was this a bug you found, or why was this necessary?